File: Win10Bluetooth/Win10Bluetooth/BluetoothDiscoverer.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Windows.Devices.Bluetooth;
using Windows.Devices.Bluetooth.Advertisement;

namespace Win10Bluetooth
{
    public class BluetoothDiscoverer
    {
        List<ulong> addresses;
        List<BluetoothLEDevice> bleDevices;
        public BluetoothDiscoverer()
        {
            addresses = new List<ulong>();
            bleDevices = new List<BluetoothLEDevice>();
        }

        public async Task<List<BluetoothLEDevice>> ScanDevices(int ms)
        {
            bleDevices.Clear();
            BluetoothLEAdvertisementWatcher watcher = new BluetoothLEAdvertisementWatcher();
            watcher.ScanningMode = BluetoothLEScanningMode.Active;
            watcher.Received += OnAdvertisementReceived;
            watcher.Start();
            await Task.Delay(ms);
            watcher.Stop();
            return bleDevices;
        }

        public bool isNewAddress(ulong newAddress)
        {
            bool isNew = true;
            foreach (ulong address in addresses)
            {
                if (address == newAddress)
                {
                    isNew = false;
                    break;
                }
            }
            return isNew;
        }

        private async void OnAdvertisementReceived(BluetoothLEAdvertisementWatcher watcher, BluetoothLEAdvertisementReceivedEventArgs eventArgs)
        {
            ulong address = eventArgs.BluetoothAddress;

            BluetoothLEDevice bleDevice = await BluetoothLEDevice.FromBluetoothAddressAsync(address);
            if (bleDevice == null)
                return;

            if (!isNewAddress(address))
                return;
            addresses.Add(address);
            bleDevices.Add(bleDevice);

        }
    }
}

File: python/BMDCameraProtocol/unused/bluetoothdiscoverer.py
import time
import asyncio

from winrt.windows.devices.bluetooth.advertisement import BluetoothLEAdvertisementWatcher, BluetoothLEAdvertisementReceivedEventArgs, BluetoothLEScanningMode
from winrt.windows.devices.bluetooth import BluetoothLEDevice

from .bmdcamera import BMDCamera
import asyncio

DEVICE_NAMES = {'bmdcamera': 'Pocket Cinema Camera 4K'}

class BluetoothDiscoverer:
    
    def __init__(self, callbackPIN=None):
        self.deviceAddresses = []
        self.ble_devices = []
        self.callbackPIN = callbackPIN

        self.watcher = BluetoothLEAdvertisementWatcher()
        self.watcher.scanning_mode = BluetoothLEScanningMode.ACTIVE
        self.watcher.add_received(self._on_advertisement_received)

    async def scan_devices(self, seconds):
        self.ble_devices = []
        self.deviceAddresses = []

        self.watcher.start()
        await asyncio.sleep(seconds)
        self.watcher.stop()

    def _on_advertisement_received(self, btAdvWatcher, btAdvEvent):
        address = btAdvEvent.bluetooth_address
        print(address)

        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        ble_device = loop.run_until_complete(self._get_device(address))
        
        if not self._isNewAddress(address):
            return
        self.deviceAddresses.append(address)

        device = self._create_device(ble_device)
        if device is not None:
            self.ble_devices.append(device)

    async def _get_device(self, address):
        device = await BluetoothLEDevice.from_bluetooth_address_async(address)
        return device

    def _create_device(self, ble_device):
        name = ble_device.name
        device = None

        if DEVICE_NAMES['bmdcamera'] in name:
            device = BMDCamera(ble_device)

        return device

    def _isNewAddress(self, address):
        isNew = True
        for dev_address in self.deviceAddresses:
            if address == dev_address:
                isNew = False
                break
        return isNew
